Document native histogram concept, send and best practices (#5956)
grafana/mimir: docs/sources/mimir/send/native-histograms/_index.md

The duration in `NativeHistogramMinResetDuration`/`withNativeResetDuration` will prohibit counter resets inside that period. Counter resets are related to the bucket limit, for more information see [Limiting the number of buckets](#limiting-the-number-of-buckets).

### Scrape and send native histograms with Prometheus

Use the latest version of Prometheus.

1. As of version 2.47, you need to enable native histograms feature via a feature flag.

   ```bash
   prometheus --enabled-feature native-histograms
   ```

1. The above flag will make Prometheus detect and scrape native histograms, but ignores classic histogram version of metrics. In case of a migration when a metric is both a classic and native histogram at the same time we recommend setting `scrape_classic_histograms` to `true` in your scrape jobs, for example:

   ```yaml
   scrape_configs:
     - job_name: myapp
       scrape_classic_histograms: true
   ```

   in your scrape jobs, to get both histogram version.

   {{% admonition type="note" %}}

>    Native histograms don't have a textual presentation at the moment on the application's `/metrics` endpoint, Prometheus actually negotiates a Protobuf protocol transfer in this case.

Reviewer: ```suggestion
   Native histograms don't have a textual presentation at the moment on the application's `/metrics` endpoint, Prometheus negotiates a Protobuf protocol transfer in this case.
```

Author: thanks

     .withClassicBuckets(0.005, 0.01, 0.025, 0,05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 25, 50, 100, Double.NaN);
     .withNativeInitialSchema(3)
     .withNativeMaxNumberOfBuckets(100)
     .withNativeResetDuration(1, TimeUnit.HOURS)
     .register();
```

{{< /code >}}

## Bucket boundary calculation

This section assumes that you are famliar with basic algebra. Native histogram bucket boundaries are calculated from an exponential formula with a base of 2.

Native histogram samples have three different kind of buckets, for any observed value the value is counted towards one kind of bucket.

- A zero bucket, which contains the count of observations whose absolute value is smaller or equal to the zero threshold.f

  [//]: # "LaTeX equation source: -threshold \\leq v \leq threshold"

  ![Zero threshold definition](zero-threshold-def.svg)

- Positive buckets, which contain the count of observations with a positive value that is greater or equal to the lower bound and smaller than the upper bound of a bucket.

  [//]: # "LaTeX equation source: {2^{\\left( 2^{-schema}\right)}}^{index} \leq v < {2^{\left( 2^{-schema}\right)}}^{index+1}"

>   ![Positive bucket definition](pos-bucket-def.svg)

Author: :blob sad: I found my sheet of paper where I wrote this done for another PR and you're right, not sure where I got this from

  [//]: # "LaTeX equation source: 2^{\\left( 2^{-schema}\right)} <= factor"

  ![Factor equation](factor-equation.svg)

  for example for factor `1.1`:

  [//]: # "Latex equation source: 2^{\\left(2^{-3}\right)}\simeq1.09<=1.1"

  ![Factor 1.1 equation](factor-1.1-equation.svg)

  Table of schema to factor:
  | _schema_ | _factor_ | | _schema_ | _factor_ |
  |----------|----------|--|----------|----------|
  | -4 | 65536 | | 3 | 1.0905 |
  | -3 | 256 | | 4 | 1.0443 |
  | -2 | 16 | | 5 | 1.0219 |
  | -1 | 4 | | 6 | 1.0109 |
  | 0 | 2 | | 7 | 1.0054 |
  | 1 | 1.4142 | | 8 | 1.0027 |
  | 2 | 1.1892 |

- Negative buckets, which contain the coint of observations with a negative value that is smaller or equal to the upper bound and larger than the lower bound of a bucket.

  [//]: # "LaTeX equation source: {-2^{\\left( 2^{-schema}\right)}}^{index+1} < v \leq {-2^{\left( 2^{-schema}\right)}}^{index}"

>   ![Negative bucket definition](neg-bucket-def.svg)

Reviewer 2: This has to be adjusted to mirror the above.

The intrumentation libraries of Prometheus have automation to keep the number of buckets down, provided that the maximum bucket number option is used (e.g. `NativeHistogramMaxBucketNumber` in GO).

Once the set maximum is exceeded, the following strategy is enacted:

1. First, if the last reset (or the creation) of the histogram is at least the minimum reset duration ago, then the whole histogram is reset to its initial state (including classic buckets). This only works if the minimum reset duration was set (`NativeHistogramMinResetDuration` in GO).

1. If less time has passed, or if the minimum reset duration is zero, no reset is performed. Instead, the zero threshold is increased sufficiently to reduce the number of buckets to or below the maximum bucket number, but not to more than the maximum zaro threashold (`NativeHistogramMaxZeroThreshold` in GO). Thus, if the threashold is at or above the maximum threashold already nothing happens at this step.

> 1. After that, if the number of buckets still exceeds maximum bucket number, the resolution of the histogram is reduced by doubling the width of all the sparse buckets (up to a growth factor between one bucket to the next of 2^(2^4) = 65536, see above in [Bucket boundary calculation](#bucket-boundary-calculation)).

Reviewer 2: Let's not say "sparse" here. :)

Author: copy paste :) https://github.com/prometheus/client_golang/blob/4c103658994b9e77dc75fbb1c78d568a8e0fa1f8/prometheus/histogram.go#L466
I'll change it.

> 1. Any increased zero threshold or reduced resolution is reset back to their original values once the minimum reset duration has passed (since the last reset or the creation of the histogram).

Reviewer 2: Which is still a bug in client_golang, BTW: https://github.com/prometheus/client_golang/issues/1248

(Doesn't have to be included here, just in case somebody asks…)

---

> description: Learn how to send metric data to Grafana Mimir.

Reviewer 3: @knylander-grafana, here is an example of content that we would need to conditionalize for _Learn how to send metric data to Grafana Mimir._ and _Learn how to send metric data to Grafana Cloud Metrics._.

How do you recommend doing that? If a solution is not feasible at this time, we could also leave off the end in favor of _Learn how to send metric data._

Reviewer 3: Similar question for [line 10](https://github.com/grafana/mimir/pull/5956/files#diff-b66a3053778bcea35b3d9b984875cb2cf907a34b597ac5ab8d8fdfd8d9bb2151R10), where we tell the reader where they are sending data to.

Author: Maybe we can simplify to say Prometheus remote write endpoint, because that can be Mimir/Cloud Metrics and many other products - not just Grafana made

```suggestion
description: Learn how to send native histograms to a Prometheus remote write endpoint.
```

keywords:
  - send metrics
  - native histogram
  - prometheus
  - grafana agent
  - instrumentation
menuTitle: Native histograms
title: Send native histograms to Mimir
weight: 100
---

# Send native histograms to Mimir

> Prometheus native histograms is a data type in the Prometheus ecosystem that enables producing, storing and querying high resolution [histogram](https://prometheus.io/docs/concepts/metric_types/#histogram) of observations.

Reviewer 3: ```suggestion
Prometheus native histograms is a data type in the Prometheus ecosystem that makes it possible to produce, store,  and query a high-resolution [histogram](https://prometheus.io/docs/concepts/metric_types/#histogram) of observations.
```

Native histograms are different from classic Prometheus histograms in a number of ways:

> 1. Native histogram bucket boundaries are calculated with a formula depending on the scale (resolution) of the native histogram and are not user defined. The calculation produces exponentially increasing bucket boundaries. For more detail see [Bucket boundary calculation](#bucket-boundary-calculation).

Reviewer 3: ```suggestion
1. Native histogram bucket boundaries are calculated by a formula that depends on the scale (resolution) of the native histogram, and are not user defined. The calculation produces exponentially increasing bucket boundaries. For details, see [Bucket boundary calculation](#bucket-boundary-calculation).
```

> 1. Native histogram bucket boundaries may change (widen) dynamically if observations result in too many buckets. For more detail see [Limiting the number of buckets](#limiting-the-number-of-buckets).

Reviewer 3: ```suggestion
* Native histogram bucket boundaries might change (widen) dynamically if the observations result in too many buckets. For details, see [Limiting the number of buckets](#limiting-the-number-of-buckets).
```

> 1. Native histogram bucket counters only count observations inside the bucket boundaries, whereas the classic histogram buckets only had an upper bound called `le` and counted all observations in the bucket and all lower buckets (cumulative).

Author: Nope, they are there to stay. Sometimes they are better even.

> 1. An instance of a native histogram metric only requires a single time series, because the buckets, sum of observations, and count of observations are stored in a single new data type - and not in separate time series using the float data type. Thus there are no `<metric>_bucket`, `<metric>_sum`, `<metric>_count` series, only `<metric>`.

Reviewer 3: ```suggestion
* An instance of a native histogram metric only requires a single time series, because the buckets, sum of observations, and the count of observations are stored in a single data type rather than in separate time series using the `float` data type. Thus, there are no `<metric>_bucket`, `<metric>_sum`, and `<metric>_count` series. There is only `<metric>`.
```
I assume you do not mean `new` data type. :) What is the new data type called?

Author: right, I'll add native histogram which is the name of the new data type

> 1. Querying native histograms via the Prometheus query language (PromQL) uses a different syntax. There are also a couple of new [functions](https://prometheus.io/docs/prometheus/latest/querying/functions/).

Reviewer 3: ```suggestion
* Querying native histograms via the Prometheus query language (PromQL) uses a different syntax. For details, see [functions](https://prometheus.io/docs/prometheus/latest/querying/functions/).
```
Future-proofing this by removing "new".

> For an introduction to native histograms, watch the following presentation: [Native Histograms in Prometheus](https://www.youtube.com/watch?v=AcmABV6NCYk).

Reviewer 3: ```suggestion
For an introduction to native histograms, watch the [Native Histograms in Prometheus](https://www.youtube.com/watch?v=AcmABV6NCYk) presentation.
```

## Pros and cons

> This section describes the pros and cons of using native histograms compared to the classic Prometheus histograms. For more information and real example, see [Prometheus Native Histograms in Production](https://www.youtube.com/watch?v=TgINvIK9SYc&t=127s) video.

Reviewer 3: ```suggestion
There are advantages and disadvantages of using native histograms compared to the classic Prometheus histograms. For more information and a real example, see the [Prometheus Native Histograms in Production](https://www.youtube.com/watch?v=TgINvIK9SYc&t=127s) video.
```

> ### Pros
> 
> 1. Simpler instrumentation. No need to think about bucket boundaries, they are automatic.
> 1. Better resolution in practice as custom bucket layouts are usually not that high resolution.
> 1. Native histograms are compatible with each other due to the automatic layout, thus they can be easily combined. Caveat: the operation might scale down an operand to lower resolution to match the other operand.

Reviewer 3: ```suggestion
### Advantages

* Simpler instrumentation: you do not need to think about bucket boundaries because they are created automatically.
* Better resolution in practice: custom bucket layouts are usually not high resolution.
* Native histograms are compatible with each other: they have an automatic layout, which makes them easy to combine. 
  {{% admonition type="note" %}}
  The operation might scale down an operand to lower resolution to match the other operand.
  {{% /admonition %}}
```
Can you test this indentation?

Author: I'll commit and check

> ### Cons
> 
> 1. Observations might be distributed in a way that is not a good fit for the exponential bucket schema, e.g. sound pressure measured in decibels, which are already logarithmic on their own.
> 1. If converting from an externally represented histogram with specific bucket boundaries, there is generally no precise match with the bucket boundaries of the native histogram and interpolation has to be applied.
> 1. There is no way to set an arbitrary bucket boundary, e.g. one that is particularly interesting for an SLO definition. Ratios of observations above or below a given threshold generally have to be estimated by interpolation, rather than being precise as it is the case for a classic histogram with a configured bucket boundary at the given threshold.
> 
> All three problems are mitigated by high resolution, which native histograms can provide at a much lower resource costs compared to classic histograms.

Reviewer 3: ```suggestion
### Cons

* Observations might be distributed in a way that is not a good fit for the exponential bucket schema, such as sound pressure measured in decibels, which are already logarithmic.
* If converting from an externally represented histogram with specific bucket boundaries, there is generally no precise match with the bucket boundaries of the native histogram, and in which case you need to use interpolation.
* There is no way to set an arbitrary bucket boundary, such as one that is particularly interesting for an SLO definition. Generally, ratios of observations above or below a given threshold have to be estimated by interpolation, rather than being precise in the case for a classic histogram with a configured bucket boundary at a given threshold.

The preceding problems are mitigated by high resolution, which native histograms can provide at a much lower resource cost compared to classic histograms.
```
Future-proofing by removing static number.

Tried to simplify and break up "Generally, ratios of observations above or below a given threshold have to be estimated by interpolation, rather than being precise in the case for a classic histogram with a configured bucket boundary at a given threshold." Please give it one more go and see if this can be broken down even further.

> ## Use native histograms

Reviewer 3: This is an extraneous section. Can it be removed?

Reviewer 3: Stopped here for now because I need to hop back to other work.

Reviewer 3: > This is an extraneous section. Can it be removed?

I meant line 46.

Author: It's meant to lead into the instrumentation bit, I'll add some text to it

   ```yaml
   scrape_configs:
     - job_name: myapp
       scrape_classic_histograms: true
   ```

   in your scrape jobs, to get both histogram version.

   {{% admonition type="note" %}}
   Native histograms don't have a textual presentation at the moment on the application's `/metrics` endpoint, thus Prometheus negotiates a Protobuf protocol transfer in this case.
   {{% /admonition %}}

1. To be able to send native histograms to a Prometheus remote write compatible receiver, for example Grafana Cloud Metrics, Mimir, etc, set `send_native_histograms` to `true` in the remote write configuration, for example:

   ```yaml
   remote_write:
     - url: http://.../api/prom/push
       send_native_histograms: true
   ```

### Scrape and send native histograms with Grafana Agent

Use the latest version of the Grafana Agent in [Flow mode](https://grafana.com/docs/agent/latest/flow/) or at least version 0.37.

1. To enable scraping native histograms you need to enable the argument `enable_protobuf_negotiation` in the `prometheus.scrape` component:

   ```
   prometheus.scrape "myapp" {
     enable_protobuf_negotiation = true
   }
   ```

> 1. The above flag will make Grafana Agent detect and scrape native histograms, but ignores classic histogram version of metrics. In case of a migration when a metric is both a classic and native histogram at the same time you need to set `scrape_classic_histogram` to `true` in your scrape jobs, for example:

Reviewer 4: ```suggestion
1. The above flag will make Grafana Agent detect and scrape native histograms, but ignores classic histogram version of metrics. In case of a migration when a metric is both a classic and native histogram at the same time you need to set `scrape_classic_histograms` to `true` in your scrape jobs, for example:
```

   {{% /admonition %}}

1. To be able to send native histograms to a Prometheus remote write compatible receiver, for example Grafana Cloud Metrics, Mimir, etc, set `send_native_histograms` argument to `true` in the `prometheus.remote_write` component, for example:

   ```
   prometheus.remote_write "mimir" {
     endpoint {
       url = "http://.../api/prom/push"
       send_native_histograms = true
     }
   }
   ```

## Migrate from classic histograms

It is perfectly possible to keep the custom bucket definition of a classic histogram and add native histogram buckets at the same time. This can ease the migration process, which can look like this in general:

1. Add native histogram definition to an existing histogram in the instrumentation.
1. Let Prometheus or Grafana Agent scrape both classic and native histogram.
1. Send native histograms to remote write - if classic histogram is scraped, it is sent by default.
1. Start modifying the recording rules, alerts, dashboards to use the new native histograms.

> 1. Once everything works, remove the custom bucket definition (`Buckets`/`classicUpperBounds`) from the instrumentation.

Reviewer 2: Or just remove the `scrape_classic_histograms: true` from the scrape config?

Author: I'll add as a final step when you got through all histograms. Because an advantage of this approach is that you can be iterative.

Author: No, you're right. I'll add as "or" , flows more natural.

     .classicUpperBounds(0.005, 0.01, 0.025, 0,05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 25, 50, 100, Double.NaN);
     .nativeInitialSchema(3)
     .nativeMaxNumberOfBuckets(100)
     .nativeResetDuration(1, TimeUnit.HOURS)
     .register();
```

{{< /code >}}

## Bucket boundary calculation

This section assumes that you are famliar with basic algebra. Native histogram bucket boundaries are calculated from an exponential formula with a base of 2.

Native histogram samples have three different kind of buckets, for any observed value the value is counted towards one kind of bucket.

- A zero bucket, which contains the count of observations whose absolute value is smaller or equal to the zero threshold.

  [//]: # "LaTeX equation source: -threshold \\leq v \leq threshold"

  ![Zero threshold definition](zero-threshold-def.svg)

- Positive buckets, which contain the count of observations with a positive value that is greater than the lower bound and less or equal to the upper bound of a bucket.

  [//]: # "LaTeX equation source: {\\left( 2^{2^{-schema}} \right)}^{index-1} < v \leq {\left( 2^{2^{-schema}}\right)}^{index}"

  ![Positive bucket definition](pos-bucket-def.svg)

>   where the _index_ can be a positive or negative integer resulting in boundaries above 1 and fractions bellow 1. The _schema_ either directly specified out of `[-4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8]` at instrumentation time or it is the largest number chosen from the list in such way that

Reviewer 2: bellow  → below

  [//]: # "LaTeX equation source: 2^{2^{-schema}} <= factor"

  ![Factor equation](factor-equation.svg)

  for example for factor `1.1`:

  [//]: # "Latex equation source: 2^{2^{-3}}\\simeq1.09<=1.1"

  ![Factor 1.1 equation](factor-1.1-equation.svg)

  Table of schema to factor:
  | _schema_ | _factor_ | | _schema_ | _factor_ |
  |----------|----------|--|----------|----------|
  | -4 | 65536 | | 3 | 1.0905 |
  | -3 | 256 | | 4 | 1.0443 |
  | -2 | 16 | | 5 | 1.0219 |
  | -1 | 4 | | 6 | 1.0109 |
  | 0 | 2 | | 7 | 1.0054 |
  | 1 | 1.4142 | | 8 | 1.0027 |
  | 2 | 1.1892 |

> - Negative buckets, which contain the count of observations with a negative value that is smaller or equal to the upper bound and larger than the lower bound of a bucket.

Reviewer 2: It has to be: "... that is smaller than the upper bound and greater than or equal to the lower bound of a bucket." The equation below is correct.

1. Loses precision if you are converting from an external representation that has stict custom bucket boundaries.
1. No way to set the bucket boundaries, which means that if a query falls between bucket boundaries, the result will be an approximation. In general this should not be a problem due to the much better resolution.

## Use native histograms

### Instrument application

TODO: should note that one can keep the old histograms
TODO: add GoLang examples. And maybe Python? Or link to examples.

### Scrape native histograms with Prometheus

TODO: should note that one can scrape both optionally

### Scrape native histograms with Grafana Agent

TODO: should note that one can scrape both optionally

### Send native histograms to Grafana Cloud with Prometheus

### Send native histograms to Grafana Cloud with Grafana Agent

## Migrate from classic histograms

### Instrument application

### Scrape metrics with Prometheus

TODO: keep scraping both

### Scape metrics with Grafana Agent

TODO: keep scraping both

## Bucket boundary calculation

Native histogram samples have three different kind of buckets, for any observed value `v` the value is counted towards one bucket.

> 1. Zero bucket, which contains the count of observations whose absolute value is smaller or equal to the zero threshold.
> 
>   [//]: # "LaTeX equation source: -ZT <= v <= ZT"
>    ![Zero threshold definition](zero-threshold-def.svg)
> 
> 1. Positive buckets, which contain the count of observations with a positive value that is greater or equal to the lower bound and smaller than the upper bound of a bucket.
> 
> 1. Negative buckets, which contain the coint of observations with a negative value that is smaller or equal to the upper bound and larger than the lower bound of a bucket.

Reviewer 3: ```suggestion
- A zero bucket, which contains the count of observations whose absolute value is smaller or equal to the zero threshold.

  [//]: # "LaTeX equation source: -ZT <= v <= ZT"
   ![Zero threshold definition](zero-threshold-def.svg)

- Positive buckets, which contain the count of observations with a positive value that is greater or equal to the lower bound and smaller than the upper bound of a bucket.

- Negative buckets, which contain the coint of observations with a negative value that is smaller or equal to the upper bound and larger than the lower bound of a bucket.
```
Ordered lists are for ordered items; use a bulleted list instead and trust that the reader can count. :)
